diff -ruN system-switch-java-1.1.5/switch_java_functions.py.in system-switch-java-1.1.5/switch_java_functions.py.in
--- system-switch-java-1.1.5/switch_java_functions.py.in	2014-03-25 15:54:00.683048705 -0400
+++ system-switch-java-1.1.5/switch_java_functions.py.in	2014-03-25 15:54:40.250233918 -0400
@@ -328,7 +328,14 @@
     vendor_arch=dest_version_vendor_arch[2].split('.')
     vendor = vendor_arch[0]
     arch = ''
-    if len(vendor_arch) > 1:
-        arch = vendor_arch[1]
+    # handle java-1.7.0-openjdk.x86_64 as well as
+    # java-1.7.0-openjdk-1.7.0.45-2.4.3.4.el7.x86_64
+    if len(dest_version_vendor_arch) > 3:
+        arch = source[source.rindex('.')+1:]
+        if arch == 'i686':
+            arch = ''
+    else:
+        if len(vendor_arch) > 1:
+            arch = vendor_arch[1]
     version = dest_version_vendor_arch[1]
     return vendor, version, arch
