--- system-switch-java-1.1.5/switch_java_tui.py	2009-08-19 20:03:01.000000000 +0200
+++ system-switch-java-1.1.5-me/switch_java_tui.py	2011-08-11 11:14:52.588453649 +0200
@@ -25,11 +25,11 @@
 
     def main(self, java_identifiers, default_java, pretty_names):
         # Initialize UI.
         screen = SnackScreen()
         text = TextboxReflowed(45, INSTRUCTION_MESSAGE)
-        list = Listbox(5, width = 20, returnExit = 1)
+        list = Listbox(5, width = 30, returnExit = 1)
         label = Label(SELECTION_MESSAGE)
         # Initialize alternatives list.
         for java in java_identifiers:
             list.append(pretty_names[java], java)
         # Show default alternative as currently selected and give it
